Implement Catalog Zones

wtoorop opened the pull request:
PR to track implementation of catalog zones in NSD.
See also issue #247 and dependent Issues.
Once #247 is done, this PR needs review and can than be merged.

Koenvh1 committed:
Add initial catalog zone config parsing and documentation as described in #250

Koenvh1 committed:
Add tpkg test for catalog zone configuration for #250

Koenvh1 committed:
Adjust checkconf test with relative path

Koenvh1 committed:
Amend nsd11 to second check

wtoorop committed:
Merge branch 'master' into features/cat-zones

Koenvh1 committed:
Fix reversed documentation for catalog and catalog-from

Koenvh1 committed:
Make config parser recognise illegal catalog zone configurations

wtoorop committed:
Merge branch 'features/cat-zones' of github.com:NLnetLabs/nsd into features/cat-zones

Koenvh1 committed:
Initialise catalog struct members

wtoorop committed:
Merge branch 'master' into features/cat-zones

wtoorop committed:
Sketch of consolidated catalog zones

+ implementation for nsd

Koenvh1 committed:
Initial implementation cat-zones-nsd

This is an intermediary update - fully untested apart from "it compiles"

wtoorop committed:
Start with processing zones

But only if read completely from file

Koenvh1 committed:
Start checks for catz requirements

wtoorop committed:
Check version owner name and content

Koenvh1 committed:
Process PTR records

wtoorop committed:
supress warnings

wtoorop committed:
Some styling suggestions

wtoorop committed:
Forgot that equal == 0

Koenvh1 committed:
Add initial removing of catalog zones

Koenvh1 committed:
Remove zones only belonging to this catalog zone

Koenvh1 committed:
Store catalog member ID

Koenvh1 committed:
Add check for COO property (untested)

Koenvh1 committed:
Improve formatting

Koenvh1 committed:
Make sure catalog zones are loaded last

Koenvh1 committed:
Fix COO property handling

Koenvh1 committed:
Do not violate 80 character limit as badly

Koenvh1 committed:
Update comments

pettai commented:
We appreciate your work on this 👍

Koenvh1 committed:
Implement group property support

jschlyter commented:
Nice work!

It would be most useful if two distinct example can be provided:

1. Primary name server use case. A config file with X zones that are published in a catalog zone for secondaries to consume.
2. Secondary name server use case. A config file with a catalog zone AXFRed from a primary that will be automatically deployed on the secondary nameserver.

I cannot really derive 1 from 2 in the man pages and it seems `catalog` and `catalog-from` are sometimes mixed up the examples.

Koenvh1 commented:
@jschlyter I will come up with the examples, but before that I am still trying hard to come up with a better way to name the catalog consumer and producer in the config, because it has been a source of confusion for many (including me).

pettai commented:
> Nice work!
> 
> It would be most useful if two distinct example can be provided:
> 
>     1. Primary name server use case. A config file with X zones that are published in a catalog zone for secondaries to consume.
> 
>     2. Secondary name server use case. A config file with a catalog zone AXFRed from a primary that will be automatically deployed on the secondary nameserver.
> 
> 
> I cannot really derive 1 from 2 in the man pages and it seems `catalog` and `catalog-from` are sometimes mixed up the examples.

Is the current code currently working? I could not get nsd to provision any zones from a loaded catalog zone with the different (catalog: / catalog-from:) options. I second the above, a working configuration example is needed.

Thx again for working on this.

pettai commented:
> @jschlyter I will come up with the examples, but before that I am still trying hard to come up with a better way to name the catalog consumer and producer in the config, because it has been a source of confusion for many (including me).

What about explicitly specifying the catalog zone(s), like @jschlyter workaround for catalog zones:
https://github.com/kirei/dnscatz (see the README).

```
catalog-zone:
     name: ...
     include-pattern: ...

```

And depending on if the include-pattern is configured to be primary or a secondary, then either generate a catalog zone, or provision the zones from the catalog.

jschlyter commented:
Some thoughts:

# Catalog Consumers

The catalog itself needs to be present on the server. One also needs information from what IP addresses and what TSIG secrets to use to fetch the zones listed in the catalog, although this might be explicit and automatically use the same values as the catalog itself. One could set up the catalog as a plain `zone` (that would allow the user to consume catalogs from local files as well as via XFR), and then have another directive that add/remove zones (using a pattern) based on catalog members.

# Catalog Producers

Just as with the consumers, declaring the catalog zone by itself as a zone and have a per zone/pattern directive that indicates that the associated zone (member) should be included in the catalog zone.


# Knot

I believe the [Knot model](https://www.knot-dns.cz/docs/3.2/html/configuration.html#catalog-zones) works quite well. Knot templates and patterns are very alike and I think a similar model would be nice for NSD as well.

Koenvh1 committed:
Process catalog zone from XFR

Koenvh1 commented:
The way it will work is the following. I am still working on adding the appropriate "hooks" in the code to make it actually load the catalog zone - previously it only loaded from file, hence it was not working :-) 

Imagine having a primary configured as
```
server:
	zonesdir: "."
	username: ""
	database: ""
	verbosity: 2
	ip-address: 0.0.0.0
	port: 55554
	pidfile: "nsd.pid"
	debug-mode: yes
        
zone:
	name: testing.catz
	zonefile: testing.catz
	notify: 127.0.0.1@55555 NOKEY # NOKEY for testing purposes only
	provide-xfr: 0.0.0.0/0 NOKEY # NOKEY for testing purposes only
```

and a secondary as
```
server:
	zonesdir: "."
	username: ""
	database: ""
	verbosity: 2
	ip-address: 0.0.0.0
	port: 55555
	pidfile: "nsd.pid"
	debug-mode: yes

pattern:
	name: "test-catalog"
	zonefile: "remote.zone"

pattern:
	name: "operator-x-foo"

zone:
	name: testing.catz
	catalog: "test-catalog"
	allow-notify: 0.0.0.0/0 NOKEY # NOKEY for testing purposes only
	request-xfr: 127.0.0.1@55554 NOKEY # NOKEY for testing purposes only
```

`testing.catz` here is the catalog zone. The `group` property matches 1-to-1 with a pattern, so for example:
```
nvxxezj.zones.testing.catz.                  0  PTR   example.net.
group.nvxxezj.zones.testing.catz.            0  TXT   (
                        "operator-x-foo" )
```
will apply the pattern "operator-x-foo" for example.net - what you configure in there is up to you. Fairly certain this is also how Knot does it.

Currently there is no support for an automatically generated/updated producer, so you will have to create that zone yourself. Whilst that is on the to-do list (#249) the consumer part has the focus currently.

jschlyter commented:
> 	name: testing.catz
> 	catalog: "test-catalog"
> 	allow-notify: 0.0.0.0/0 NOKEY # NOKEY for testing purposes only
> 	request-xfr: 127.0.0.1@55554 NOKEY # NOKEY for testing purposes only

"catalog" is really the default pattern for the catalog, correct? But can be overridden by any group property.

Koenvh1 committed:
Serve zone added from catalog zone

Koenvh1 committed:
Add member_id and catalog name to zone added from catalog

Koenvh1 committed:
Add pattern for catzone if it does not already exist

Koenvh1 committed:
Add code to handle patterns

Koenvh1 committed:
Fix group property with tasks

Koenvh1 committed:
Clean up redundant code

Koenvh1 committed:
Make check COO a separate task

Koenvh1 committed:
Use NSD region for catname

Koenvh1 commented:
Documentation is still lacking (saving the best for last :-) ), but I think the core functionality (from the consumer perspective) is now implemented. I have tried to come up with some good automated tests, but as all the tricky parts stem from the network interaction between primary and secondary (for example the COO property), I have not been able to come up with something proper (yet). Something that is also untested is interoperability with other implementations.

Short rundown of what has been added:
* Zones are added based on a catalog zone, by default with a pattern with the name of the catalog zone (e.g. `catalog.testing.catz.`)
* The group property changes the pattern of a zone to the pattern with the name of the group property
* The COO property works once 1) NSD is aware of the COO property in the old zone and 2) the zone that wants to take over the ownership gets reloaded. On startup this may mean that the zone is added based on the old catalog zone, but this will resolve itself over time.
* Zones added through a catalog zone are not added to the zone list. 

You will find that it tends to do a couple of loops through all the processes to ensure that every process within NSD is aware of all the changes in order not to run into inconsistent state issues. This is not necessarily pretty, nor does it necessarily make it easy to reason about what it is doing, but it seemed the safest way to ensure that every process agrees on the current state.

Koenvh1 committed:
Update documentation on catalog zones

Koenvh1 committed:
Resolve gcc warnings

Koenvh1 committed:
Remove commented out code

Koenvh1 committed:
Remove free statement

Koenvh1 committed:
Add catzone test

Koenvh1 commented:
The man page for nsd.conf now contains all the main information about how catalog zones currently work. There is a test that tests the happy path for catalog zones along with some bug fixes that I found along the way.

Koenvh1 committed:
Attempt to fix CI pipeline

pettai commented:
Hi @Koenvh1,

I've built your latest git version of NSD with catz-support, but the catalog zone I'm fetching from remote doesn't seem to trigger any automatically generated secondaries zones. The catz-zone is very simple and doesn't have any groups. A `nsd-control write` generates the following file:

```
; zone catz-test.catalog written by NSD 4.6.2 on Sun Jul 16 14:09:46 2023
; received update to serial 1686895545 at 2023-07-16T14:06:48 from XX:YY:ZZ::4 TSIG verified with key zzzz-key
$ORIGIN catalog.
catz-test	0	IN	SOA	invalid. invalid. (
		1686895545 3600 600 2147483646 0 )
	0	IN	NS	invalid.
$ORIGIN catz-test.catalog.
version	0	IN	TXT	"2"
$ORIGIN zones.catz-test.catalog.
1a3e5acaxxxxxxxx	0	IN	PTR	catztest1.se.
2d3c2381yyyyyyyy	0	IN	PTR	catztest2.se.
```

The gist of my NSD conf looks like this:

```
# DNSOPS CATZ test
server:
        debug-mode: yes

pattern:
        name: "catz-test-pattern"
        zonefile: "/var/lib/nsd/zones/%s"
        # allow-notify: XX:YY:ZZ::4 zzzz-key
        # request-xfr:  XX:YY:ZZ::4 zzzz-key

zone:
        name: "catz-test.catalog"
        # include-pattern: "catz-test-pattern"
        catalog: "catz-test-pattern"
        allow-notify: XX:YY:ZZ::4 zzzz-key
        request-xfr:  XX:YY:ZZ::4 zzzz-key
```

It doesn't work w/o`include-pattern` vs. local zone config (just like your example config above)
Perhaps the catalog zone data I'm getting isn't working as intended with NSD? 
(I believe it's generated by bind or a knot server, @jschlyter probably knows)

And there are no errors in the NSD log file either...

wcawijngaards commented on dbaccess.c:
```diff
@@ -33,6 +33,9 @@
 #include "ixfr.h"
 #include "ixfrcreate.h"
 
+/* TODO: Maybe eventually include "cat-zones.h" */
```
This TODO item seems already dealt with.

```suggestion
```

wcawijngaards commented on cat-zones-nsd.c:
```diff
@@ -0,0 +1,172 @@
+
+	DEBUG(DEBUG_CATZ, 1, 
+	(LOG_INFO, "Task created for catalog %s: %s", catname, zname));
+	task_new_add_catzone(udb, last_task, zname, pname, catname, dname_to_string(member_id, NULL), 0);
```
If this function just creates a new task, which it sends over the pipe, then the allocated strings are leaked and not freed? The zname and catname that have been region_strdupped. The catname could be been used for the `patopt->pname`, and then it needs to be allocated at the end of the function, but I do not see why the string are allocated like that. Perhaps copy them into a temporary buffer of some sort during the action, because they seem to be leaked.

wcawijngaards commented on cat-zones-nsd.c:
```diff
@@ -0,0 +1,172 @@
+				break;
+			}
+
+			if (dname->label_count == 
```
The equality on the labelcount could be awkward if the new zone name has several labels, so a new zone with more than one label.

wcawijngaards commented on cat-zones-nsd.c:
```diff
@@ -0,0 +1,172 @@
+				DEBUG(DEBUG_CATZ, 1, 
+				(LOG_INFO, "Group property discovered"));
+
+				task_new_apply_pattern(udb, last_task, dname_to_string(dname, NULL), (const char*)(rdata_atom_data(rr->rdatas[0]) + 1));
```
Check the length of the rrdata, that the wire text contains data.

wcawijngaards commented on cat-zones-nsd.c:
```diff
@@ -0,0 +1,172 @@
+
+				const dname_type* member_id = rr->owner->dname;
+
+				catz_add_zone(
```
If this is the only place where `catz_add_zone` is called, there does not seem to be a reason to pass `NULL` and cast `nsd` to `void*`.

wcawijngaards commented on cat-zones-nsd.c:
```diff
@@ -0,0 +1,172 @@
+	DEBUG(DEBUG_CATZ, 1, 
+	(LOG_INFO, "Task created for catalog %s: %s", catname, zname));
+	task_new_add_catzone(udb, last_task, zname, pname, catname, dname_to_string(member_id, NULL), 0);
+	return 1;
```
The return value seems to be unused.

wcawijngaards commented on cat-zones-nsd.c:
```diff
@@ -0,0 +1,172 @@
+		}
+	}
+
+	return -1;
```
The return value seems to be unused.

wcawijngaards commented on dbaccess.c:
```diff
@@ -674,9 +682,38 @@ void namedb_check_zonefiles(struct nsd* nsd, struct nsd_options* opt,
 	RBTREE_FOR(zo, struct zone_options*, opt->zone_options) {
-		namedb_check_zonefile(nsd, taskudb, last_task, zo);
-		if(nsd->signal_hint_shutdown) break;
+			zone_type* zone;
```
Maybe it is github's diff display, but perhaps it should be one tab to the left.
```suggestion
		zone_type* zone;
```

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+
+	pattern_options_type* patopt;
+
+	log_msg(LOG_INFO, "addcatzone task %s %s %s %s", zname, pname, catname, member_id);
```
The debug output could be with like DEBUG_IPC or DEBUG_CATZ. That is already in the next line.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+	patopt = pattern_options_find(nsd->options, pname);
+	if (!patopt) {
+		patopt = pattern_options_create(nsd->region);
+		patopt->pname = pname;
```
The variable `pname` here refers to the task element, and that is going to be deleted when it has been processed. But then `pattern_options_add_modify` allocates, both a new pattern and the name. But then it could be that the `patopt` is allocated but unused at the end of the function.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+		return;
+	}
+	z->zonestatid = (unsigned)task->yesno;
+	z->from_catalog = (char*)catname;
```
When the `catname` variable, that refers to the task element, gets deallocated when the task element is removed after it is processed. Then the `from_catalog` is no longer pointing to a valid string.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+
+	log_msg(LOG_INFO, "addcatzone task %s %s %s %s", zname, pname, catname, member_id);
+	DEBUG(DEBUG_IPC,1, (LOG_INFO, "addcatzone task %s %s %s %s", zname, pname, catname, member_id));
+	zdname = dname_parse(nsd->db->region, zname);
```
It looks like find_or_create_zone does not actually store the name passed to it, and that means `zdname` could be allocated but not in use at the end of the function. Except the code path where it is recycled.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+
+	zone_type* t;
+
+	log_msg(LOG_INFO, "checkcoo task %s %s %s", zname, catname, member_id);
```
Debug output could be with like DEBUG_IPC or DEBUG_CATZ.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+	zone_type* t;
+
+	log_msg(LOG_INFO, "checkcoo task %s %s %s", zname, catname, member_id);
+	zdname = dname_parse(nsd->region, zname);
```
The `zdname` variable is allocated, but no longer used at the end of the function.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+		}
+		cz = namedb_find_zone(
+			nsd->db, 
+			dname_parse(
```
The return value of `dname_parse` here is allocated, but not freed, and unused at the end of the function.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+
+		DEBUG(DEBUG_CATZ, 1, 
+			(LOG_INFO, 
+			"Found existing zone belong to a catalog zone %s", 
```
Spelling.
```suggestion
			"Found existing zone belonging to catalog zone %s", 
```

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+		}
+
+		if (coo_correct) {
+			t->from_catalog = (char*)catname;
```
The contents of `catname` gets deallocated after the function ends, so `from_catalog` is then invalid.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+
+		if (coo_correct) {
+			t->from_catalog = (char*)catname;
+			t->catalog_member_id = (dname_type*)member_id;
```
The contents of `member_id` gets deallocated after the function ends, so `catalog_member_id` is then invalid. Because the task item is removed, some time after the tasks have been processed.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+			t->catalog_member_id = (dname_type*)member_id;
+			return;
+		} else {
+			return;
```
The else branch is empty, because the function ends after it anyway.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+
+	zdname = dname_parse(nsd->region, member_id);
+
+	log_msg(LOG_INFO, "applypattern task %s %s", member_id, pname);
```
Debug output should not end up in the final version.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+	zone_type* gz = NULL;
+	struct zone_options* zo;
+
+	zdname = dname_parse(nsd->region, member_id);
```
The `zdname` contents is not deallocated at the end of the function.

wcawijngaards commented on difffile.c:
```diff
@@ -2025,6 +2103,204 @@ task_process_add_zone(struct nsd* nsd, udb_base* udb, udb_ptr* last_task,
+			const char* ncatalog_member_id = region_strdup(nsd->region, dname_to_string(gz->catalog_member_id, NULL));
+
+			task_new_del_zone(udb, last_task, gz->apex->dname);
+			task_new_add_catzone(
```
Since the `task_new_add_catzone` function copies the variables, they are unused and still allocated at the end of the function, `nzname`, `nfrom_catalog` and `ncatalog_member_id`.

wcawijngaards commented on nsd.conf.5.in:
```diff
@@ -996,10 +996,18 @@ Feed updated zone to verifier over standard input. Default is
 to not use a specific timeout. Default is verifier\-timeout from
 .B verify:\fR.
+.TP
+.B catalog:\fR <pattern>
```
The `catalog-from` option is not documented.

wcawijngaards commented on cat-zones-nsd.c:
```diff
@@ -0,0 +1,172 @@
+	// Re-add all zones coming from a catalog
+
+	const char* catname = region_strdup(nsd->region, 
+		dname_to_string(zone->apex->dname, NULL));
```
The configure option `--disable-radix-tree` fails for this, the dname element is accessed with `domain_dname` or with `domain_to_string`.

wcawijngaards commented on server.c:
```diff
@@ -2352,6 +2353,15 @@ server_reload(struct nsd *nsd, region_type* server_region, netio_type* netio,
+		zone->opts->pattern->catalog && 
+		zone->is_updated) {
+			DEBUG(DEBUG_CATZ, 1, (LOG_INFO, "start catalog consumption"));
+			nsd_catalog_consumer_process(nsd, zone, 
```
This performs a full delete and add, also after IXFR or partial change of the catalog zone. The delete uses delete zone and that also removes the zone data from the zones that continue to be present. But, on the other hand, this code makes catalog support work, so it is fine to add.

k0ekk0ek replied:
This code should be moved up? There's also the case with verification, catalog zones are included for the checks, but we process it whether it's considered good or bad. I think we have to decide whether catalog zones should be included for verification (might be useful?, maybe make it configurable?) and update behavior here accordingly. So, if we decide to allow for that, we probably shouldn't process the contents of the "broken" zone as per the specification? There's a couple of quirks here I'm not sure about yet.

wcawijngaards commented on tpkg/checkconf.tdir/checkconf.check2:
The difference here is just the nsd12 file output. The check2 file contains more differences from check that do not seem to be system specific output issues. It seems that the check2 file does not work properly. The commit does not change this, but the error output from nsd12 should really be present, I believe.

Koenvh1 replied:
Then I think I misunderstood the difference between `check` and `check2`. I thought they were system specific output differences, as the test seems to check whether either check or check2 matches the output.

Regardless, the error will be added

Koenvh1 replied:
I discussed it briefly with @gthess - given that the `catalog-from` option currently does nothing, I have removed it. This also removed the need for `nsd12` as that merely checks for a potential conflict between `catalog` and `catalog-from`. Once/If the publisher side of catalog zones gets added we can re-add it.

wcawijngaards commented on tpkg/catzone.tdir/catzone-secondary.conf:
```diff
@@ -0,0 +1,57 @@
+
+pattern:
+	name: "testing.catz."
+	# zonefile: "example.com.remote.zone"
```
Should this be `"%s.remote.zone"` as that allows the pattern to be used for different zones, or maybe the pattern is only there to test one zone. The zonefile record is commented out anyway, and would only be written after the zonetransfer and zonefile write action, if a filename that does not exist yet is chosen.

wcawijngaards commented on tpkg/catzone.tdir/catzone.conf:
```diff
@@ -0,0 +1,16 @@
+	chroot: ""
+	pidfile: nsd.pid
+	logfile: nsd.log
+	database: nsd.db
```
Perhaps a test with `database: ""` is nicer, we advise that value for users, and consider obsoleting it. It also causes different zone loading and unloading code to be used. So it could maybe make a difference in that aspect.

wcawijngaards commented on tpkg/catzone.tdir/catzone.conf:
```diff
@@ -0,0 +1,16 @@
+
+zone:
+	name: example.com
+	zonefile: cname_soa.zone
```
The zonefile name seems to be a reference to the test that was copied from. Perhaps it is an unneeded leftover.

wcawijngaards commented on tpkg/catzone.tdir/catzone.test:
```diff
@@ -0,0 +1,37 @@
+# #-- cname_soa.test --#
```
In other test scripts this refers to the test name.
```suggestion
# #-- catzone.test --#
```

wcawijngaards commented on tpkg/catzone.tdir/catzone.test:
```diff
@@ -0,0 +1,37 @@
+if dig @127.0.0.1 -p 55555 example.com SOA | grep "dns1.example.com"; then
+	echo OK
+else
+	cat nsd.log
```
The file nsd.log is not used, perhaps with files `cat nsd-primary.log` and `cat nsd-secondary.log`.

wcawijngaards commented on tpkg/catzone.tdir/catzone.test:
```diff
@@ -0,0 +1,37 @@
+# if grep "zone example.com file cname_soa.zone read with 3 errors" nsd.log; then
+# 	echo "detected the errors OK"
+# else
+# 	cat nsd.log
```
The nsd.log file is not in use.

wcawijngaards commented on tpkg/catzone.tdir/catzone.test:
```diff
@@ -0,0 +1,37 @@
+kill_pid `cat nsd-primary.pid`
+kill_pid `cat nsd-secondary.pid`
+
+# if grep "zone example.com file cname_soa.zone read with 3 errors" nsd.log; then
```
This comment seems to be a leftover. Maybe it should be removed.

wcawijngaards commented on tpkg/catzone.tdir/example.zone:
The file `example.zone` seems to be unused in this test.

Koenvh1 replied:
The `example.zone`` file is the zone that gets loaded on the secondary from the catalog zone served by the primary.

wcawijngaards commented on xfrd.c:
```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
+		patopt = pattern_options_find(xfrd->nsd->options, pattern);
+		if (!patopt) {
+			patopt = pattern_options_create(xfrd->region);
+			patopt->pname = pattern;
```
The variable pattern is part of the task, that is deleted after the function ends. But pattern is added to the options. It should have contents that is allocated.

wcawijngaards commented on xfrd.c:
```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
+			zopt = zone_list_zone_insert(xfrd->nsd->options, zname, pattern, 0, 0);
+		} else {
+			log_msg(LOG_INFO, "catzone already added %s", zname);
+			return;
```
At this point, the memory allocated in `dname_parse(xfrd->region, zname)` is not freed, and the function exits.

wcawijngaards commented on xfrd.c:
```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
+		if (!zopt) {
+			zopt = zone_list_zone_insert(xfrd->nsd->options, zname, pattern, 0, 0);
+		} else {
+			log_msg(LOG_INFO, "catzone already added %s", zname);
```
The output print should use VERBOSITY or the DEBUG for CATZ output.

wcawijngaards commented on xfrd.c:
```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
+		} 
+		zopt = zone_options_find(xfrd->nsd->options, dname_parse(xfrd->region, zname));
+		if (!zopt) {
+			zopt = zone_list_zone_insert(xfrd->nsd->options, zname, pattern, 0, 0);
```
This adds it to the zone.list file, instead of `nsd_options_insert_zone` that just adds it to the list of zones in memory in the options structure. At position 0,0 in the file, would this mess up the zone.list bookkeeping, and is the catalog zone not separate from that.

Koenvh1 replied:
`options.h` has as comment for `zone_list_zone_insert`: `/* create zonelist entry, do not insert in file (called by _add) */` - should I change that as well?

wcawijngaards replied:
The `zone_list_add` function is used by `remote.c` where it adds the zone that needs to be configured additionally, and it adds it to the zonelist part of the config, in memory, and also in the zonelist part of the config on disk in the zonelist file. The zone structure then also needs to be added into the options in memory for the other processes in NSD, this is done by the `zone_list_zone_insert` function that is called from `difffile.c` when the task to edit that contents arrives from the tasklist. It would then not add it to the zonelist file, because one process is modifying the zonelist file, that is the xfrd process that does that. For the zone structure that is added in catalog zones, I am not sure what it is that needs to be done, hence I was asking as well, perhaps it needs to add this in a different way. It should not modify the zonelist file on disk, and then also not change the 'zonelist file' data structure in memory?

Koenvh1 replied:
What I discussed with @wtoorop (in person) was that adding to the zone.list file itself is probably not necessary, as the catalog zone acts as kind of a zone list in and of itself. Let's see whether I can do that using `nsd_options_insert_zone` instead.

wcawijngaards commented on xfrd.c:
```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
+			log_msg(LOG_INFO, "catzone already added %s", zname);
+			return;
+		}
+		task_new_add_catzone(xfrd->nsd->task[xfrd->nsd->mytask],
```
When xfrd receives a catzone addition task as a taskresult from reload, it then makes a new task for the next reload to add the catzone? Is that intentionally the work order, or a loop?

Koenvh1 replied:
Because this is called from xfrd, the result of the new task ends up at difffile rather than back here, so there is no loop (or there should not be). Admittedly this part still confuses me as well.

wcawijngaards commented on xfrd.c:
```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
+		/* add to xfrd - notify (for master and slaves) */
+		init_notify_send(xfrd->notify_zones, xfrd->region, zopt);
+		/* add to xfrd - slave */
+		xfrd_init_slave_zone(xfrd, zopt);
```
It may be nice to guard this call to `xfrd_init_slave_zone` with an if on `zone_is_slave(zopt)` , so that a catalog zone could add primary zones. If the zone is primary depends on the config for it and this is determined by the config pattern that is used, and since that can be set, it seems possible.

wcawijngaards commented on xfrd.c:
```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
+		/* add to xfrd - notify (for master and slaves) */
+		init_notify_send(xfrd->notify_zones, xfrd->region, zopt);
+		/* add to xfrd - slave */
+		xfrd_init_slave_zone(xfrd, zopt);
```
It may be nice to guard this call to `xfrd_init_slave_zone` with an if on `zone_is_slave(zopt)` , so that a catalog zone could add primary zones. If the zone is primary depends on the config for it and this is determined by the config pattern that is used, and since that can be set, it seems possible.

wcawijngaards replied:
This item seems to be a duplicate comment, that was generated as a copy from another. I'll close it.

wcawijngaards reviewed (approved):
The code looks to be a nice addition and the catalog zone processing is a nice feature!

Koenvh1 committed:
Remove catalog-from

Koenvh1 committed:
Use separate region for catzone

Koenvh1 committed:
Remove unnecessary logging

Koenvh1 committed:
Fix catzone test

Koenvh1 committed:
Ensure pattern name is not cleared

Koenvh1 committed:
Add `zone_is_slave` check

pettai commented:
**UPDATED**

After building the latest version NSD crashes after loading the catalog zone _from disk (file)_. 
nsd-checkconf doesn't complain, and the configuration is the same as stated in the comment above (no changes).

```
...
Aug 16 16:08:49 bygg-u2204 nsd[19710]: listen on ip-address ::1@53 (udp) with server(s): *
Aug 16 16:08:49 bygg-u2204 nsd[19710]: listen on ip-address ::1@53 (tcp) with server(s): *
Aug 16 16:08:49 bygg-u2204 nsd[19710]: listen on ip-address 127.0.0.1@53 (udp) with server(s): *
Aug 16 16:08:49 bygg-u2204 nsd[19710]: listen on ip-address 127.0.0.1@53 (tcp) with server(s): *
Aug 16 16:08:49 bygg-u2204 nsd[19710]: file rotation on (null) enabled
Aug 16 16:08:49 bygg-u2204 nsd[19710]: dropped user privileges, run as nsd
Aug 16 16:08:49 bygg-u2204 nsd[19710]: xfrd pre-startup
Aug 16 16:08:49 bygg-u2204 nsd[19710]: xfrd: adding catz-test.catalog zone
Aug 16 16:08:49 bygg-u2204 nsd[19710]: xfrd zone catz-test.catalog is activated, state 1
Aug 16 16:08:49 bygg-u2204 nsd[19710]: xfrd: started server 1 secondary zones
Aug 16 16:08:49 bygg-u2204 nsd[19710]: task procsync /tmp/nsd-xfr-19710/nsd.19710.task.1 size 288
Aug 16 16:08:49 bygg-u2204 nsd[19711]: delete zone visit catz-test.catalog.
Aug 16 16:08:49 bygg-u2204 nsd[19711]: axfrdel: recyclebin holds 0 bytes
Aug 16 16:08:49 bygg-u2204 nsd[19711]: memory: 26 objects (26 small/0 large), 1536 bytes allocated (43 wasted) in 1 chunks, 0 cleanups, 0 in recyclebin 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
Aug 16 16:08:49 bygg-u2204 nsd[19711]: zone catz-test.catalog read with success
Aug 16 16:08:49 bygg-u2204 nsd[19711]: Catz version TXT
Aug 16 16:08:49 bygg-u2204 nsd[19711]: Catz version is 2
Aug 16 16:08:49 bygg-u2204 nsd[19711]: 1a3e5aca65819ae2.zones
Aug 16 16:08:49 bygg-u2204 nsd[19711]: Task created for catalog catz-test.catalog.: catztest1.se.
Aug 16 16:08:49 bygg-u2204 nsd[19711]: add task addcatzone catztest1.se. catz-test.catalog. catz-test.catalog. 1a3e5aca65819ae2.zones.catz-test.catalog.
Aug 16 16:08:49 bygg-u2204 kernel: [84389.527356] nsd: main[19711]: segfault at 20 ip 00005575b86842b5 sp 00007fff89844330 error 4 in nsd[5575b865a000+7f000]
Aug 16 16:08:49 bygg-u2204 kernel: [84389.527381] Code: 00 e8 0f 74 fd ff 66 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 40 00 f3 0f 1e fa 41 57 41 56 41 55 41 54 55 48 89 f5 53 48 83 ec 28 <48> 8b 5e 20 48 8d 71 11 48 89 7c 24 10 4c 8b 0b 89 54 24 1c 48 89
Aug 16 16:08:49 bygg-u2204 nsd[19710]: did not get start signal from main
```

Koenvh1 committed:
Add check to see that group property is not empty

Koenvh1 committed:
Remove left-overs from "generic" catzone implementation

Koenvh1 committed:
Make zdname recyclable, only insert zone in memory

Koenvh1 committed:
Ensure variable lifetime is appropriate

Koenvh1 committed:
Fix lifetime issues

Koenvh1 committed:
Recycle space properly

pettai commented:
Latest build gets a bit longer, the catalog zones are added according to the logs

```
Aug 16 06:37:33 bygg-u2204 nsd[16733]: zone catz-test.catalog. received update to serial 1686895545 at 2023-08-16T06:37:33 from XX:YY:ZZ::4 TSIG verified with key zzzz-key of 249 bytes in 0.000803 seconds
Aug 16 06:37:33 bygg-u2204 nsd[16732]: now adding catzone catztest1.se. catz-test.catalog. catz-test.catalog. 1a3e5aca65819ae2.zones.catz-test.catalog.
Aug 16 06:37:33 bygg-u2204 nsd[16732]: now adding catzone catztest2.se. catz-test.catalog. catz-test.catalog. 2d3c238127656416.zones.catz-test.catalog.
```

But `nsd-control` doesn't present which the zones are:

```
bygg-u2204:/home/pettai# nsd-control zonestatus
zone:	catz-test.catalog
	state: ok
	served-serial: "1686895545 since 2023-08-06T01:19:47"
	commit-serial: "1686895545 since 2023-08-16T06:37:33"
	wait: "3360 sec between attempts"
zone:
	pattern: catz-test.catalog.
	state: master
zone:
	pattern: catz-test.catalog.
	state: master 
```

Btw. the pattern name is actually `catz-test-pattern`, but `nsd-control` show the catalog zone name here

```
pattern:
        name: "catz-test-pattern"
        zonefile: "/var/lib/nsd/zones/%s"

zone:
        name: "catz-test.catalog"
        catalog: "catz-test-pattern"
```

And the don't get loaded correctly (according to EDE) and queries doesn't work:

```
bygg-u2204:/home/pettai# dig catztest1.se @localhost

; <<>> DiG 9.18.12-0ubuntu0.22.04.2-Ubuntu <<>> catztest1.se @localhost
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 30436
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; EDE: 14 (Not Ready): (Zone is configured but not loaded)
;; QUESTION SECTION:
;ri.se.				IN	A

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(localhost) (UDP)
;; WHEN: Wed Aug 16 06:40:12 UTC 2023
;; MSG SIZE  rcvd: 73
```

pettai commented:
Setting the name of the pattern stanza to the exact name of the catalog zone itself (including a ending "." dot) finally makes NSD fetch and load the zones from the catalog correctly 👍 (for a catalog test zone without groups configured https://github.com/NLnetLabs/nsd/pull/261#issuecomment-1637113588)

This could be explained more clearly in the man page please, but Ideally, it shouldn't be needed or just an invisible default 

```
pattern:
        name: "catz-test.catalog."
        zonefile: "/var/lib/nsd/zones/%s"

zone:
        name: "catz-test.catalog"
        catalog: "catz-test.catalog."
```

Now yields:

```
# nsd-control zonestatus
zone:	catz-test.catalog
	state: ok
	served-serial: "1686895545 since 2023-08-21T15:01:13"
	commit-serial: "1686895545 since 2023-08-21T15:41:28"
	wait: "3354 sec between attempts"
zone:
	pattern: catz-test.catalog.
	state: ok
	served-serial: "1950 since 2023-08-21T15:41:28"
	commit-serial: "1950 since 2023-08-21T15:41:28"
	wait: "816 sec between attempts"
zone:
	pattern: catz-test.catalog.
	state: ok
	served-serial: "2004024163 since 2023-08-21T15:41:28"
	commit-serial: "2004024163 since 2023-08-21T15:41:28"
	wait: "895 sec between attempts"
```

Writing out the individual zone names in `nsd-control zonestatus` output would be an nice improvement 👍  

But still, after doing a `nsd-control write` and a restart of NSD, NSD still crashes upon loading the catalog zone https://github.com/NLnetLabs/nsd/pull/261#issuecomment-1667293301 though...

Koenvh1 committed:
Make sure zonename does not get lost

k0ekk0ek commented on install-sh:
Is there a good reason we're updating this file in this PR?

k0ekk0ek commented on namedb.h:
```diff
@@ -150,6 +150,8 @@ struct zone
 	unsigned     is_skipped : 1; /* subsequent zone updates are skipped */
 	unsigned     is_checked : 1; /* zone already verified */
 	unsigned     is_bad : 1; /* zone failed verification */
+	char*        from_catalog; /* zone originates from catalog zone */
+	dname_type*  catalog_member_id; /* member id in the catalog zone */
```
Is it wise to make `catalog_member_id` a pointer to the actual domain(?) We can then retrieve the `catalog` to which it belongs via that reference. Otherwise we'll probably need to do separate lookups every time, which can be costly(?) (Just a thought, don't know if it actually makes sense before I finish the review, might introduce a problem when removing zones...)

Koenvh1 committed:
Keep a list of changes from the catalog zone, apply them afterwards

Current behaviour is that all changes are applied whilst the catalog zone is processed. This means that for a zone with a group property, the following tasks are created in order: 1) remove zone, 2) add zone with default pattern, 3) remove zone, 4) add zone with pattern from group property.

This changes that to one operation

Koenvh1 committed:
Do not abuse line limits as much

Koenvh1 committed:
Resolve GCC warning

Koenvh1 committed:
Fix static analyzer warning

k0ekk0ek commented on options.h:
```diff
@@ -293,6 +293,7 @@ struct pattern_options {
 	uint8_t verifier_feed_zone_is_default;
 	int32_t verifier_timeout;
 	uint8_t verifier_timeout_is_default;
+	const char* catalog;
```
This seems off(?) I'd expect this to be a boolean value(?) Either a zone is a catalog zone or not? Based on whether the server is a primary or secondary, that'd have different implications. Currently it almost seems to specify which catalog to use, but that behavior is implemented by `include-pattern`. The pattern itself would dictate where to get the catalog zone from? The boolean value would merely indicate post-processing is necessary. Then when reading the catalog zone file, we'd mark other zones as necessary automatically?

k0ekk0ek commented:
@wtoorop, @wcawijngaards, @pettai: @Koenvh1 and I will be collaborating on finishing catalog zones. Closing this PR in favor of #294, which we'll treat as a feature branch and add updates as we go along.

pettai commented:
Ok, I will start building & testing from that branch source then 👍
